text/trace/0168-sampling-propagation.md

```diff
@@ -0,0 +1,368 @@
+100% sampling probability (i.e., `s=0` yielding `p=1`).
+
+### Behavior of the `AlwaysOff` Sampler
+
+The `AlwaysOff` Sampler behaves the same as `TraceIDRatioBased` with
+zero probability (i.e., `p=63`, `s` undefined).
+
+## Worked 3-bit example
+
+The behavior of these tables can be verified by hand using a smaller
+example.  The following table shows how these equations work where
+`r`, `p`, and `s` are limited to 3 bits.
+
+Values of `p`, which have the same encoded value and interpretation as
+for the proposed `log_head_adjusted_count` field of OTEP 170, would be
+interpreted as follows:
+
+| `p` value | Adjusted count |
+| -----     | -----          |
+| 0         | Unknown        |
+| 1         | 1              |
+| 2         | 2              |
+| 3         | 4              |
+| 4         | 8              |
+| 5         | 16             |
+| 6         | 32             |
+| 7         | 0              |
+
+Note there are only 6 non-zero, non-unknown values for the adjusted
+count. Thus there are six defined values of `r` and `s`.  The
+following table shows `r` and the corresponding selection probability,
+along with the calculated adjusted count for each `s`:
+
+| `r` value | `r` selection probability | `s=0` | `s=1` | `s=2` | `s=4` | `s=5` | `s=6` |
+| --        | --                        | --    | --    | --    | --    | --    | --    |
+| 0         | 1/2                       | 1     | 0     | 0     | 0     | 0     | 0     |
+| 1         | 1/4                       | 1     | 2     | 0     | 0     | 0     | 0     |
+| 2         | 1/8                       | 1     | 2     | 4     | 0     | 0     | 0     |
+| 3         | 1/16                      | 1     | 2     | 4     | 8     | 0     | 0     |
+| 4         | 1/32                      | 1     | 2     | 4     | 8     | 16    | 0     |
+| 5         | 1/32                      | 1     | 2     | 4     | 8     | 16    | 32    |
+
+Notice that the sum of `r` selection probability times adjusted count
+in each of the `s=*` columns equals 1.  For example, in the `s=5`
+column we have `0*1/2 + 0*1/4 + 0*1/8 + 0*1/16 + 16*1/32 + 16*1/32 =
+16/32 + 16/32 = 1`.  In the `s=2` column we have `0*1/2 + 0*1/4 +
+4*1/8 + 4*1/16 + 4*1/32 + 4*1/32 = 4/8 + 4/16 + 4/32 + 4/32 = 1/2 +
+1/4 + 1/8 + 1/8 = 1`.  We conclude that when `r` is chosen with the
+given probabilities, any choice of `s` produces one expected span.
+
+## Summary
+
+The following table summarizes how the three Sampler cases behave with
+respect to the incoming and outgoing values for `p`, `r`, and
+`sampled`:
+
+| Sampler                | Incoming `r` | Incoming `p` | Incoming `sampled` | Outgoing `r`    | Outgoing `p`   | Outgoing `sampled` |
+| --                     | --           | --           | --                 | --              | --             | --                 |
+| Parent                 | unused       | expected     | respected          | passed through  | passed through | passed through     |
```

Comment:
Parent-based sampler should not pass through inconsistent values of `r`, `p`, and `sampled` (see my comment below).

Reply:
I rewrote this as "checked and passed through", then added a table of range checks that should be applied, then documented the invariant as you stated it. I want there to be a special case for `sampled` is true and `p=63` to allow intentional sampling with zero adjusted count. I would be glad to prototype a composite sampler that applies either the probability sampler's decision (resulting in p in [1..62]) or a non-probability sampler's decision (effectively `p=63`, saying "collect this for some reason but don't count it") to show this in action.